### ISSUE TITLE
Facilitate space-efficient throw-away clones of dataset handles

### ISSUE THREAD

**Issue by mih** [datalad/datalad#1]
If I have a big dataset and I want to do a number of analyses with multiple users in a shared computing environment, it would be nice to be able to do that in a way that has minimal impact on storage demands.

In my concrete case, I have about 100GB of raw input data that is currently copied for each clone of the dataset handle. Of course I can avoid that by manually hard/soft linking the relevant files, and only later inject potential results of derived data into the original annex. However, this breaks the workflow with dataset handles.

**Comment by hanke** [comment]
I am reopening this one again. Here is a scenario that I feel is poorly supported right now -- maybe that is the local optimum already, but maybe it could be done in a different way.

Imagine a webserver exposing a git annex repo (A) via http. I have a collaborator that needs to get access to additional unpublished parts of a new revision of the dataset in (A). My approach is to go to the webserver and

1. `git clone --shared` (A) into (B)
2. merge the branch with the new bits from a remote into (B)
3. `git annex get .`

The last step will hardlink files in the annex of (A) when possible and copy the rest from the remote of (B) into (B). Now I should be be able to give the HTTP URL of (B) to my collaborator and git annex should do the rest for them.

However, the problem is that, most likely, they cannot clone (B) via http. Git assumes with `--shared` that it can obtain objects from the `.git` of (A) when cloning (B) via the same access method. Example:

(B) is cloned locally from /media/data/orig from (A) and is exposed via http://BASEURL/t002
(C) is cloned from (B) via http://BASEURL/t002

This fails at the git clone stage, because git requests objects from (A) via http://BASEURL/media/data/orig (remote URL and local path concatenated).

In order to fix this, the webserver needs to get an Alias definition that makes such URLs valid.

Long story short: It would be good to have a switch that enables git-annex annex hard-linking without the implications `git clone --shared`. Actually, the ideal solution seems to be the combination of git's behavior with `clone --local` and what git annex is doing now with `clone --shared`.

@joeyh Does this make sense, or am I approaching it wrong?

**Comment by joeyh** [comment]
Michael Hanke wrote:
> However, the problem is that, most likely, they cannot clone (B) via http. Git
> assumes with --shared that it can obtain objects from the .git of (A) when
> cloning (B) via the same access method.

Interesting point.

I wonder if using smart http transport avoids this problem? My guess is
it will, because then git is running on the server and can follow the
alternates.

> Long story short: It would be good to have a switch that enables git-annex
> annex hard-linking without the implications git clone --shared. Actually, the
> ideal solution seems to be the combination of git's behavior with clone --local
> and what git annex is doing now with clone --shared.

Well, you can clone --local and then git config annex.hardlink true
and get the configuration you want.

Since --local is the default, git-annex obviously cannot set the
generally unsafe annex.hardlink by default.

-- 
see shy jo

**Comment by hanke** [comment]
Thanks @joeyh!

If there are no unintended side effects of setting the flag manually, this seems like the way to go. I'll try that out in practice.

**Comment by joeyh** [comment]
Michael Hanke wrote:
> If there are no unintended side effects of setting the flag manually, this
> seems like the way to go. I'll try that out in practice.

It's perfectly fine to set the flag manually, just beware of the
cautions. In particular you should probably mark the repository that has
annex.hardlink set as untrusted.

-- 
see shy jo

**Comment by hanke** [comment]
I finally go around to try this and it works as advertised. Closing.

**Comment by hanke** [comment]
Ok, now I feel stupid. I forgot to test the actual remote operation -- and there it doesn't work, but the error is new. Here is the protocol:

on the server
````
% git clone --local datasets/forrest_gump task002
% cd task002
% git update-server-info
% git annex untrust here
% git config annex.hardlink true
% git annex get stimulus
get stimulus/task001/annotations/german_audio_description.csv (from origin...) ok
get stimulus/task001/annotations/scenes.csv (from origin...) ok
get stimulus/task001/generate/generate_melt_cmds.py (from origin...) ok
get stimulus/task001/psychopy/buttons.csv (from origin...) ok
get stimulus/task001/psychopy/forrest_gump.psyexp (from origin...) ok
get stimulus/task001/psychopy/segment_cfg.csv (from origin...) ok
(Recording state in git...)
````

on the client

```
% git clone http://URL/task002/.git stupid`
% git annex init
% git annex get stimulus/task001/annotations/scenes.csv
get stimulus/task001/annotations/scenes.csv (not available) 
  No other repository is known to contain the file.
failed
git-annex: get: 1 failed
````

Whenever I don't configure hard linking it works like charm, though. I checked that the file in question is actually hard-linked and it is. It seems that the clone on the client side doesn't even try to get the file -- apache shows no access attempt. Here is what annex says with debug on:

````
% git annex get stimulus/task001/annotations/scenes.csv -d
[2015-02-27 17:42:11 CET] read: git ["--git-dir=/tmp/stupid/.git","--work-tree=/tmp/stupid","ls-files","--cached","-z","--","stimulus/task001/annotations/scenes.csv"]
get stimulus/task001/annotations/scenes.csv [2015-02-27 17:42:11 CET] read: git ["--git-dir=/tmp/stupid/.git","--work-tree=/tmp/stupid","show-ref","git-annex"]
[2015-02-27 17:42:11 CET] read: git ["--git-dir=/tmp/stupid/.git","--work-tree=/tmp/stupid","show-ref","--hash","refs/heads/git-annex"]
[2015-02-27 17:42:11 CET] read: git ["--git-dir=/tmp/stupid/.git","--work-tree=/tmp/stupid","log","refs/heads/git-annex..3801da6521898aad97428cf1689c861919697b02","-n1","--pretty=%H"]
[2015-02-27 17:42:11 CET] chat: git ["--git-dir=/tmp/stupid/.git","--work-tree=/tmp/stupid","cat-file","--batch"]
(not available) 
  No other repository is known to contain the file.
failed
git-annex: get: 1 failed
````

@joeyh Do you know what is going on. As you said it should work, I am assuming that I made some mistake, but if you think this is a bug, I can file a report against git annex.

Thanks!

**Comment by yarikoptic** [comment]
it seems that it is simply because
1. original repository is N/A (that git annex initrepo NAME type=git  should help)
2. I guess running 'git update-server-info' (or just enabling post-update hook) should fixup list of available files

**Comment by joeyh** [comment]
Right, need to run git update-server-info after git-annex get so that the git-annex branch update will be seen over http

But, no existing git hook script runs in this situation. git-annex should provide one.

**Comment by mih** [comment]
Closing this historic issue -- it has served us well and long enough!
